File: CourseProjects/CourseProjects/JavaEclipseProjects/DesignPatternsExamples/src/patterns/visitor/XMLFormatVisitor.java
package patterns.visitor;

import java.io.*;

public class XMLFormatVisitor implements Visitor {
	private BufferedWriter output;

	public XMLFormatVisitor(Writer w) {
		if(w instanceof BufferedWriter) {
			output = (BufferedWriter)w;
		} else {
			output = new BufferedWriter(w);
		}	
	}
	public void visitDocumentStart(Document d) {
		write("<document>");
	}
	public void visitDocumentEnd() {
		write("<document>");
		closeOutput();
	}
	public void visitModuleStart(Module m) {
		write("<module>");
	}
	public void visitModuleEnd() {
		write("</module>");
	}
	public void visitParagraphStart(Paragraph p) {
		write("<paragraph>");
		write(p.toString());
	}
	public void visitParagraphEnd(){
		write("</paragraph>");
	}
	public void visitSectionStart(Section s) {
		write("<section>");
	}
	public void visitSectionEnd() {
		write("</section>");
	}
	public void visitTitleStart(Title title) {
		write("<title>");
		write(title.toString());
	}
	public void visitTitleEnd(){
		write("</title>");
	}
	private void write(String text) {
		try {
			output.write(text);
			output.newLine();
		} catch(IOException ex) {
			System.out.println(ex);
		}
	}
	private void closeOutput() {
		try {
			output.close();
		} catch(IOException ex) {
			System.out.println(ex);
		}
	}
}

File: CourseProjects/CourseProjects/VisualStudioProjects/DesignPatternsExamples/Visitor/XMLFormatVisitor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Visitor {
    internal class XMLFormatVisitor : IVisitor {
        private readonly StreamWriter output;

        public XMLFormatVisitor(FileStream fs) {
            output = new StreamWriter(fs);
        }

        public void VisitDocumentStart(Document d) {
            Write("<document>");
        }

        public void VisitDocumentEnd() {
            Write("</document>");
            CloseOutput();
        }

        public void VisitModuleStart(Module m) {
            Write("<module>");
        }

        public void VisitModuleEnd() {
            Write("</module>");
        }

        public void VisitParagraphStart(Paragraph p) {
            Write("<paragraph>");
            Write(p.ToString());
        }

        public void VisitParagraphEnd() {
            Write("</paragraph>");
        }

        public void VisitSectionStart(Section s) {
            Write("<section>");
        }

        public void VisitSectionEnd() {
            Write("</section>");
        }

        public void VisitTitleStart(Title title) {
            Write("<title>");
            Write(title.ToString());
        }

        public void VisitTitleEnd() {
            Write("</title>");
        }

        private void Write(string text) {
            output.Write(text);
            output.WriteLine();
        }

        private void CloseOutput() {
            output.Close();
        }
    }
}